Moving from string(zone) to endpointGroupInfo when mapping NEGs, and populate subnet values for nodes in zoneGetter and NEG syncers.

sawsa307 opened the pull request:
* Use endpointGroupInfo instead of string(zone) when mapping NEGs since now endpoints shall be partitioned based on the zone and the subnet they are in.
* Nodes in non-default subnets will be filtered when enableMultiSubnetCluster=true and enableMultiSubnetClusterPhase1=false. When enableMultiSubnetClusterPhase1=true, we will include endpoints
  and nodes from non-default subnets.

gauravkghildiyal commented:
/assign

sawsa307 commented:
/retest

sawsa307 commented:
cc @mmamczur

mmamczur commented on pkg/neg/types/interfaces.go:
```diff
@@ -81,11 +81,16 @@ type NegSyncerManager interface {
+
+type EndpointInfo struct {
+	Zone   string
+	Subnet string
```
so the empty Subnet value means the default subnet?
What should be the content of this field for the non default subnets?

sawsa307 replied:
> so the empty Subnet value means the default subnet?

I'm open to suggestions here. For now I leave it as empty string since most of nodes in default subnets doesn't have this label. A different approach is we can manually pipe the default subnet value so it is not empty

swetharepakula replied:
I think both situations need to be handled. We need to understand the empty string means the default subnet and handle that some nodes might have empty and some might have the subnet name.

mmamczur commented on pkg/neg/types/interfaces.go:
```diff
@@ -81,11 +81,16 @@ type NegSyncerManager interface {
 	ValidateEndpoints(endpointData []EndpointsData, endpointPodMap EndpointPodMap, endpointsExcludedInCalculation int) error
 }
+
+type EndpointInfo struct {
```
'EndpointInfo' suggests it holds information about an endpoint - in our case a pod for l7 or node for l4, while the struct contents are data that map 1:1 to the NetworkEndpoint**Group**. Maybe it could be `EndpointGroupInfo` or `GroupZoneAndSubnet` or something like that

sawsa307 replied:
Sounds good, updated to `EndpointGroupInfo`.

gauravkghildiyal commented:
I haven't taken a look, but I thought this was a draft?

sawsa307 commented:
> I haven't taken a look, but I thought this was a draft?

Yes it is still a draft. I converted to "ready to review" so we can get some result on the `pull-ingress-gce-e2e` and `pull-ingress-gce-test`.

sawsa307 commented:
/hold This is still a draft.

gauravkghildiyal commented:
> > I haven't taken a look, but I thought this was a draft?
> 
> Yes it is still a draft. I converted to "ready to review" so we can get some result on the `pull-ingress-gce-e2e` and `pull-ingress-gce-test`.

Oh gotcha! Yep all I want is for you to save time on not trying to perfect it without deciding which path you want to take.

sawsa307 commented:
/assign @swetharepakula

sawsa307 commented:
/assign @mmamczur

swetharepakula commented on pkg/neg/syncers/endpoints_calculator_test.go:
```diff
@@ -62,8 +62,8 @@ func TestLocalGetEndpointSet(t *testing.T) {
 			endpointsData: negtypes.EndpointsDataFromEndpointSlices(getDefaultEndpointSlices()),
```
do we have tests that cover both cases of nodes. Some with the empty string as the subnet and some with the default subnet name

sawsa307 replied:
No that's not needed. We will make sure the subnet value is not empty. This enhancement is introduced in the last PR.

If a node/endpoint is in the default subnet, the subnet field in its `EndpointGroupInfo` should contain the value of the defaultSubnet.

swetharepakula reviewed (commented):
Please expand in the commit messages to provide more context such as "moving from string (zone) to the endpoint group info in preparation for creating NEGs per subnet".

The last commit (`Use defaultSubnet as subnet value for endpoints in default subnet.`), I don't understand what the goal of this commit is? It is just passing in the default subnet. Do we need this? In most of these places you will have to change the code again when we start pulling the info for the other subnets.

mmamczur commented on pkg/neg/syncers/transaction.go:
```diff
@@ -498,12 +505,12 @@ func (s *transactionSyncer) syncNetworkEndpoints(addEndpoints, removeEndpoints m
 				go s.attachNetworkEndpoints(zone, batch)
```
so I guess here you also need to pass down the subnet

sawsa307 replied:
Added a todo, thanks

mmamczur commented on pkg/neg/controller.go:
```diff
@@ -170,6 +171,15 @@ func NewController(
 
+	var defaultSubnet string
+	if flags.F.EnableMultiSubnetClusterPhase1 {
+		defaultSubnet, err = utils.KeyName(defaultSubnetURL)
```
here you have access to cloud, so you could just do `cloud.SubnetworkURL()` here or even in the manager, unless you prefer this for testing purposes

sawsa307 replied:
Updated, thanks

sawsa307 commented on pkg/neg/syncers/transaction.go:
```diff
@@ -678,7 +680,8 @@ func (s *transactionSyncer) commitPods(endpointMap map[string]negtypes.NetworkEn
 			zoneEndpointMap[endpoint] = podName
 		}
-		s.reflector.CommitPods(s.NegSyncerKey, s.NegSyncerKey.NegName, zone, zoneEndpointMap)
+		// Only zone is needed because NEGs from non-default subnet has different names.
```
Updated.

sawsa307 reviewed (commented):
> Do we need this? In most of these places you will have to change the code again when we start pulling the info for the other subnets.

This is to make sure we have a uniform way to present endpoints in the default subnet, instead of having empty string and default subnet value representing the same thing.
There are only two places we generate `EndpointGroupInfo`. One is during endpoint calculation, which we can get the subnet value from the node. The other one is during `retrieveExistingZoneNetworkEndpointMap`, which should only contains endpoints from the default subnet. It will require update anyways because we don't currently create NEGs from non-default subnet.

sawsa307 commented on pkg/neg/types/interfaces.go:
```diff
@@ -81,11 +81,16 @@ type NegSyncerManager interface {
 	ValidateEndpoints(endpointData []EndpointsData, endpointPodMap EndpointPodMap, endpointsExcludedInCalculation int) error
 }
+
+type EndpointGroupInfo struct {
```
Done.

sawsa307 commented:
/retest

mmamczur commented:
lgtm from my side, we need this to be able to do endpoint calculator changes

gauravkghildiyal commented on pkg/neg/syncers/subsets.go:
```diff
@@ -193,7 +193,7 @@ func getSubsetPerZone(nodesPerZone map[string][]*v1.Node, totalLimit int, svcID
 				ip = utils.GetNodePrimaryIP(node, logger)
 			}
-			result[zone.Name].Insert(negtypes.NetworkEndpoint{Node: node.Name, IP: ip})
+			result[negtypes.EndpointGroupInfo{Zone: zone.Name, Subnet: defaultSubnet}].Insert(negtypes.NetworkEndpoint{Node: node.Name, IP: ip})
```
There's a potential bug here -- this part seems to be assuming that the Node is in the defaultSubnet, and I don't think such a constraint is enforced anywhere.

If you've already discussed this with @mmamczur and fixing this is something outside the scope of this PR, let's atleast add a comment here so that it can be fixed shortly after by the L4 team in a followup PR.

gauravkghildiyal commented on pkg/neg/manager.go:
```diff
@@ -227,6 +245,7 @@ func (manager *syncerManager) EnsureSyncers(namespace, name string, newPorts neg
 				manager.syncerMetrics,
 				&portInfo.NetworkInfo,
 				portInfo.L4LBType,
+				manager.defaultSubnet,
```
Why do we need a separate `defaultSubnet` when we are already passing NetworkInfo? Why can't we resuse the networkInfo? (Similar comment for the TransactionSyncer construction)

gauravkghildiyal commented on pkg/neg/syncers/endpoints_calculator.go:
```diff
@@ -49,20 +49,24 @@ type LocalL4EndpointsCalculator struct {
 }
 
-func NewLocalL4EndpointsCalculator(nodeLister listers.NodeLister, zoneGetter *zonegetter.ZoneGetter, svcId string, logger klog.Logger, networkInfo *network.NetworkInfo, lbType types.L4LBType) *LocalL4EndpointsCalculator {
+func NewLocalL4EndpointsCalculator(nodeLister listers.NodeLister, zoneGetter *zonegetter.ZoneGetter, svcId string, logger klog.Logger, networkInfo *network.NetworkInfo, lbType types.L4LBType, defaultSubnet string) *LocalL4EndpointsCalculator {
```
Why do we need a separate defaultSubnet, when networkInfo has the same information?

sawsa307 replied:
This is to ensure we only parse the information once in manager.go. The alternative I can think of is we parse the defaultSubnet from defaultSubnetURL in `NewTransactionSyncer` and `GetEndpointsCalculator`.

sawsa307 replied:
Just saw the other comment. Now it makes more sense to generate on the fly. Let me update this.

gauravkghildiyal commented on pkg/neg/syncers/transaction.go:
```diff
@@ -156,6 +159,15 @@ func NewTransactionSyncer(
 
 	logger := log.WithName("Syncer").WithValues("service", klog.KRef(negSyncerKey.Namespace, negSyncerKey.Name), "negName", negSyncerKey.NegName)
 
+	var defaultSubnet string
+	var err error
+	if flags.F.EnableMultiSubnetClusterPhase1 {
+		defaultSubnet, err = utils.KeyName(networkInfo.SubnetworkURL)
+		if err != nil {
+			logger.Error(err, "Errored getting default subnet from NetworkInfo")
+		}
+	}
```
What do we gain from pre-calculating the defaultSubnet during the TransactionSyncer object construction? Why can't we calculate it when we want? Doesn't the current code lead to a situation where you are logging an error, but still continuing assuming everything is just fine? (I know sometimes these things are unavoidable, but in this case this to me seems to hint towards the fact that perhaps this is not the right place to do this)

sawsa307 replied:
Okay yes that makes sense. Generate on need sounds like a better solution. Let me update this, thank you!

gauravkghildiyal reviewed (commented):
Thanks for addressing the comments @sawsa307!

Besides the two comments that I have, the rest seems reasonable to me. 

I will not block the PR on the following, but I somehow feel the naming `EndpointGroupInfo` might be overly vague. Given the idea with the struct is to define what is the granularity for each NEG (i.e. For each `EndpointGroupInfo`, we have a unique NEG), so how does something like `NEGLocation` sound? Anyone else has any thoughts here? There's also the option of `ZoneSubnet` -- I know we likely are avoiding this to be flexible for some potential future use case but perhaps this vagueness might not be fully justified here. 

Either way, we can likely do this renaming in a followup PR to unblock the rest of the feature development.

sawsa307 commented:
> Thanks for addressing the comments @sawsa307!

Besides the two comments that I have, the rest seems reasonable to me.

I will not block the PR on the following, but I somehow feel the naming EndpointGroupInfo might be overly vague. Given the idea with the struct is to define what is the granularity for each NEG (i.e. For each EndpointGroupInfo, we have a unique NEG), so how does something like NEGLocation sound? The idea being, for each location (NEGLocation), I have a NEG there. Anyone else has any thoughts here? There's also the option of ZoneSubnet -- I know we likely are avoiding this to be flexible for some potential future use case but perhaps this vagueness might not be fully justified here.

Either way, we can likely do this renaming in a followup PR to unblock the rest of the feature development.

Thank you so much Gaurav for reviewing this!
I'm fine with either naming(NEGLocation/EndpointGroupInfo). Definitely would like more opinions on this.
There is one another change included since the previous review. I discussed with @swetharepakula and we think we should try to simplify the logic in the zoneGetter regarding `defaultSubnet`, given the two msc phases have different behaviors towards endpoints in non-default subnets.
The change being made is: Now everything will based on the field `onlyIncludeDefaultSubnetNodes` in zoneGetter. If it is true, we will filter out all nodes in non-default subnets. Otherwise, we will populate the correct subnet value for the node.

gauravkghildiyal commented on pkg/neg/syncers/transaction.go:
```diff
@@ -259,7 +259,16 @@ func (s *transactionSyncer) syncInternalImpl() error {
-	currentMap, currentPodLabelMap, err := retrieveExistingZoneNetworkEndpointMap(s.NegSyncerKey.NegName, s.zoneGetter, s.cloud, s.NegSyncerKey.GetAPIVersion(), s.endpointsCalculator.Mode(), s.enableDualStackNEG, s.logger)
+	var defaultSubnet string
+	var err error
+	if !flags.F.EnableMultiSubnetCluster {
```
If `flags.F.EnableMultiSubnetCluster = true`:

- defaultSubnet for currentMap = ""
- In targetMap, from zone getter, defaultSubnet received would depend on whether the Node has the label so both "" and "<SUBNET_NAME>" values are possible (depending on whether the Node had a label or not)

Is it guaranteed that for the phase when `flags.F.EnableMultiSubnetCluster = true`, then the Node Labels for subnet will never be set?

sawsa307 replied:
Sorry for the confusion. Just updated the code again since the logic is very confusing.
Right now `defaultSubnet` should not be empty in any case. In the currentMap, we will use defaultSubnet = `the value of the default subnet`, and in targetMap and zoneGetter, `defaultSubnet` received will be based on the node label. If the label is non-empty, the value is the actual subnet value. If the label is empty or doesn't exist, we will return the actual value of the default subnet as well. Please let me know if there is more confusion on this, thank you!

gauravkghildiyal commented on pkg/utils/zonegetter/zone_getter.go:
```diff
@@ -331,6 +339,26 @@ func getZone(node *api_v1.Node) (string, error) {
+// If a node doesn't have this label, or the label value is empty, it means
+// this node is in the defaultSubnet and we will parse the subnet from the
+// defaultSubnetURL.
+func getSubnet(node *api_v1.Node, defaultSubnetURL string) (string, error) {
```
This seems almost similar to `isNodeInDefaultSubnet`, is there no way to combine these two?

sawsa307 replied:
That's a good point. Refactored, thanks!

gauravkghildiyal reviewed (commented):
Almost there.

---

From the commit message:

```
Populate subnet values for nodes in zoneGetter and NEG syncers.
* Nodes in non-default subnets will be filtered when
  enableMultiSubnetCluster=true. This behavior will be overriden by
  enableMultiSubnetClusterPhase1 flag, since we will include non-default
  subnet nodes then.
```

Does this mean when `enableMultiSubnetClusterPhase1=FALSE`, we will start considering nodes from non-default subnet?

sawsa307 commented on pkg/neg/syncers/transaction.go:
```diff
@@ -259,7 +259,12 @@ func (s *transactionSyncer) syncInternalImpl() error {
-	currentMap, currentPodLabelMap, err := retrieveExistingZoneNetworkEndpointMap(s.NegSyncerKey.NegName, s.zoneGetter, s.cloud, s.NegSyncerKey.GetAPIVersion(), s.endpointsCalculator.Mode(), s.enableDualStackNEG, s.logger)
+	defaultSubnet, err := utils.KeyName(s.networkInfo.SubnetworkURL)
+	if err != nil {
+		s.logger.Error(err, "Errored getting default subnet from NetworkInfo")
```
Updated.

sawsa307 commented on pkg/neg/syncers/subsets.go:
```diff
@@ -173,13 +173,20 @@ func getSubsetPerZone(nodesPerZone map[string][]*v1.Node, totalLimit int, svcID
 
+	defaultSubnet, err := utils.KeyName(networkInfo.SubnetworkURL)
+	if err != nil {
+		logger.Error(err, "Errored getting default subnet from NetworkInfo")
```
Updated.

sawsa307 commented:
> Almost there.
> 
> From the commit message:
> 
> ```
> Populate subnet values for nodes in zoneGetter and NEG syncers.
> * Nodes in non-default subnets will be filtered when
>   enableMultiSubnetCluster=true. This behavior will be overriden by
>   enableMultiSubnetClusterPhase1 flag, since we will include non-default
>   subnet nodes then.
> ```
> 
> Does this mean when `enableMultiSubnetClusterPhase1=FALSE`, we will start considering nodes from non-default subnet?

Updated the commit message so it correctly reflects the behavior.

gauravkghildiyal commented:
> > Almost there.
> > From the commit message:
> > ```
> > Populate subnet values for nodes in zoneGetter and NEG syncers.
> > * Nodes in non-default subnets will be filtered when
> >   enableMultiSubnetCluster=true. This behavior will be overriden by
> >   enableMultiSubnetClusterPhase1 flag, since we will include non-default
> >   subnet nodes then.
> > ```
> > 
> > Does this mean when `enableMultiSubnetClusterPhase1=FALSE`, we will start considering nodes from non-default subnet?
> 
> Updated the commit message so it correctly reflects the behavior.

My concern is whether the behavior correctly matches the commit message; isn't this contradictory (from zone getter)?

```
onlyIncludeDefaultSubnetNodes: flags.F.EnableMultiSubnetCluster && flags.F.EnableMultiSubnetClusterPhase1,
```

sawsa307 commented:
> Body

Oh yes the second flag should be negated. Pushed up another commit.

gauravkghildiyal commented:
Thanks for the patience. This LGTM. Once the other related changes merge, I hope we can thoroughly test this out manually.

If people don't have a strong preference, I'd suggest we rename `EndpointGroupInfo` to `NEGLocation` in a followup PR -- I hope that should not be as intense as this given it's a simple rename.

/lgtm
/hold (In case you are waiting for additional reviews)

gauravkghildiyal commented:
/label tide/merge-method-squash

gauravkghildiyal commented:
/lgtm
/unlabel tide/merge-method-squash

gauravkghildiyal commented:
/remove-label tide/merge-method-squash

sawsa307 commented:
Trying to squash and rebase the code based on swetharepakula@'s request.

gauravkghildiyal commented:
/lgtm

sawsa307 commented:
/unhold

swetharepakula commented on pkg/utils/zonegetter/zone_getter_test.go:
```diff
@@ -199,62 +200,80 @@ func TestListNodesMultipleSubnets(t *testing.T) {
 				}
+			}
 
+			for _, enableMultiSubnetClusterPhase1 := range []bool{true, false} {
```
I think it is a great idea to re-use the test cases, but I would put these into its own t.Run as the earlier part of the test and this part are not connected.

swetharepakula commented on pkg/neg/syncers/subsets.go:
```diff
@@ -173,13 +173,21 @@ func getSubsetPerZone(nodesPerZone map[string][]*v1.Node, totalLimit int, svcID
 	// Sort zones in increasing order of node count.
 	zoneList := sortZones(nodesPerZone)
 
+	defaultSubnet, err := utils.KeyName(networkInfo.SubnetworkURL)
```
made a comment on a previous commit. we should just store this and calculate it at initialization. if there is a problem better to throw an error at startup because it will alway cause an error

swetharepakula commented on pkg/neg/syncers/transaction_test.go:
```diff
@@ -1332,9 +1343,11 @@ func TestTransactionSyncerWithNegCR(t *testing.T) {
 	testNegType := negtypes.VmIpPortEndpointType
 	flags.F.EnableMultiSubnetClusterPhase1 = true
+	defer func() {
+		flags.F.EnableMultiSubnetClusterPhase1 = false
```
this should store the previous value and set it back then assuming that the flag will be false
